android/src/main/java/com/oney/WebRTCModule/WebRTCDevicesManager.java

```diff
@@ -0,0 +1,247 @@
+            enumerateDevicesArray.pushMap(params);
+        }else {
+            WritableMap paramsBuiltIn = this.createWritableMap(DeviceType.BUILT_IN_MICROPHONE, "Built in microphone", DeviceKind.AUDIO_INPUT.getKind());
+            enumerateDevicesArray.pushMap(paramsBuiltIn);
+
+            WritableMap paramsBuiltInSpeaker = this.createWritableMap(DeviceType.BUILT_IN_MICROPHONE_SPEAKER, "Built in speaker microphone", DeviceKind.AUDIO_INPUT.getKind());
+            enumerateDevicesArray.pushMap(paramsBuiltInSpeaker);
+        }
+
+        boolean isBluetoothHeadsetPlugged = Arrays.stream(audioInputDevices).anyMatch(device -> device.getType() == AudioDeviceInfo.TYPE_BLUETOOTH_SCO);
+        if(isBluetoothHeadsetPlugged){
+            WritableMap params = this.createWritableMap(DeviceType.BLUETOOTH, "Bluetooth microphone", DeviceKind.AUDIO_INPUT.getKind());
+            enumerateDevicesArray.pushMap(params);
+        }
+    }
+
+    private void fillAudioOutputDevices(WritableArray enumerateDevicesArray){
+        AudioDeviceInfo[] audioOutputDevices = this.audioManager.getDevices(AudioManager.GET_DEVICES_OUTPUTS);
+
+        boolean isWiredHeadsetPlugged = Arrays.stream(audioOutputDevices).anyMatch(
+                device -> device.getType() == AudioDeviceInfo.TYPE_WIRED_HEADSET ||
+                        device.getType() == AudioDeviceInfo.TYPE_WIRED_HEADPHONES
+        );
+
+        WritableMap params = null;
+        // At Android, if the Wired headset is plugged, we are unable to choose to use the built-in microphone. We can only choose to use the bluetooth or the default.
+        if(isWiredHeadsetPlugged){
+            params = this.createWritableMap(DeviceType.HEADSET, "Wired headset", DeviceKind.AUDIO_OUTPUT.getKind());
+        }else {
+            params = this.createWritableMap(DeviceType.EARPIECE, "Phone Earpiece", DeviceKind.AUDIO_OUTPUT.getKind());
+        }
+        enumerateDevicesArray.pushMap(params);
+
+        //speaker
+        params = this.createWritableMap(DeviceType.SPEAKER, "Speaker", DeviceKind.AUDIO_OUTPUT.getKind());
+        enumerateDevicesArray.pushMap(params);
+
+        boolean isBluetoothHeadsetPlugged = Arrays.stream(audioOutputDevices).anyMatch(device -> device.getType() == AudioDeviceInfo.TYPE_BLUETOOTH_SCO);
+        if(isBluetoothHeadsetPlugged){
+            params = this.createWritableMap(DeviceType.BLUETOOTH, "Bluetooth", DeviceKind.AUDIO_OUTPUT.getKind());
+            enumerateDevicesArray.pushMap(params);
+        }
+    }
+
+    private WritableMap createWritableMap(DeviceType deviceType, String label, String kind){
+        WritableMap audioMap = Arguments.createMap();
+        audioMap.putString("deviceId", deviceType.toString());
+        audioMap.putString("groupId", "");
+        audioMap.putString("label", label);
+        audioMap.putString("kind", kind);
+        if(deviceType.audioRoute != null){
+            audioMap.putInt("audioRoute", deviceType.audioRoute.getValue());
```

Comment:
since we're defining this field, what if instead of an int we actually used a human-readable string? could make it easier for customers to learn and understand.

Reply:
I can do that, sure.

Reply:
Fixed.